tools/codegen/src/generator/languages/rust_lazy_reader/generator.rs

```diff
@@ -0,0 +1,445 @@
+use super::{ident_new, LazyReaderGenerator};
+/// ## Terminology
+/// - field, existing in "struct", "table"
+/// - field name, name of field
+/// - field type, type of field
+/// - field type name, name of field type
+/// - item, existing in "dynvec", "union" and "option"
+/// - item name, there is no such thing, the item is without name
+/// - item type, type of item
+/// - item type name, name of item type
+/// - type name, item type name or field type name
+/// - raw name, a struct name without "Type", same as name
+/// - type category, field type or item type are classified into 3 categories, see TypeCategory.
+///   Different category has different implementation.
+/// - C/Rust transformed name: transformed from "field type name" or "item type name",
+///   according to their type category, e.g. uint32_t, mol2_cursor_t, XXXType
+/// - name, used for class name only
+/// - class type name, name with "Type" suffix, e.g. SampleTableType
+/// - common array, the set of Array, FixVec and DynVec, they share method like "len", "get"
+/// - common table, the set of Table, Struct, they share same method like ".t->XXX"
+use crate::ast::{self, HasName, *};
+use proc_macro2::{Literal, TokenStream};
+use quote::quote;
+use std::io;
+
+impl LazyReaderGenerator for ast::Union {
+    fn gen_rust<W: io::Write>(&self, output: &mut W) -> io::Result<()> {
+        let name = ident_new(self.name());
+
+        let q = quote! {
+            pub struct #name {
+                pub cursor: Cursor,
+            }
+
+            impl From<Cursor> for #name {
+                fn from(cursor: Cursor) -> Self {
+                    Self { cursor }
+                }
+            }
+
+            impl #name {
+                pub fn item_id(&self) -> Result<usize, Error> {
+                    let item = self.cursor.union_unpack()?;
+                    Ok(item.item_id)
+                }
+            }
+        };
+        writeln!(output, "{}", q)?;
+
+        for item in self.items() {
+            let item_type_name = item.typ().name();
+            let item_type_name = ident_new(&format!("as_{}", item_type_name.to_lowercase()));
+            let (transformed_name, tc) = get_rust_type_category(item.typ());
+            let convert_code = tc.gen_convert_code();
+            let q = quote! {
+                impl #name {
+                    pub fn #item_type_name(&self) -> Result<#transformed_name, Error> {
```

Comment:
I think we need to add an union id verification here, otherwise an union with the same variant type could result in the wrong type being converted without an error being reported.